### PR TITLE
Implement correct bind clause rules and refactor to use new parser.Match util.

### LINKED ISSUE #162: "bad cast" when using alias as argument
Input:
```
fs default() {
        myAlias
        download "."
}

fs foo(option::run opts) {
        image "busybox"
        run "touch foo" with opts
}

fs bar() {
        foo option::run {
                mount scratch "/tmp" as myAlias
                dir "/tmp"
        }
}
```

```
$ ./hlb run ./test.hlb
#1 compiling [default]
#1 0.000 github.com/openllb/hlb/codegen.(*CodeGen).EmitFilesystemChainStmt.func1
#1 0.000        /src/github.com/openllb/hlb/codegen/chain.go:121
#1 0.000 github.com/openllb/hlb/codegen.(*CodeGen).EmitChainStmt.func1
#1 0.000        /src/github.com/openllb/hlb/codegen/chain.go:44
#1 0.000 github.com/openllb/hlb/codegen.(*CodeGen).EmitBlock
#1 0.000        /src/github.com/openllb/hlb/codegen/codegen.go:314
#1 0.000 github.com/openllb/hlb/codegen.(*CodeGen).EmitFilesystemBlock
#1 0.000        /src/github.com/openllb/hlb/codegen/codegen.go:340
#1 0.000 github.com/openllb/hlb/codegen.(*CodeGen).EmitFuncDecl
#1 0.000        /src/github.com/openllb/hlb/codegen/decl.go:37
#1 0.000 github.com/openllb/hlb/codegen.(*CodeGen).Generate
#1 0.000        /src/github.com/openllb/hlb/codegen/codegen.go:138
#1 0.000 github.com/openllb/hlb.Compile.func1
#1 0.000        /src/github.com/openllb/hlb/hlb.go:91
#1 0.000 github.com/openllb/hlb/solver.(*progressUI).Write.func2.1
#1 0.000        /src/github.com/openllb/hlb/solver/progress.go:172
#1 0.000 github.com/openllb/hlb/solver.write
#1 0.000        /src/github.com/openllb/hlb/solver/progress.go:196
#1 0.000 github.com/openllb/hlb/solver.(*progressUI).Write.func2
#1 0.000        /src/github.com/openllb/hlb/solver/progress.go:171
#1 0.000 golang.org/x/sync/errgroup.(*Group).Go.func1
#1 0.000        /src/github.com/openllb/hlb/vendor/golang.org/x/sync/errgroup/errgroup.go:57
#1 0.000 runtime.goexit
#1 0.000        /usr/local/go/src/runtime/asm_amd64.s:1373
#1 0.000 Caused by: ./test.hlb:2:2: bad cast
#1 ERROR: ./test.hlb:2:2: bad cast
------
 > compiling [default]:
#1 0.000 Caused by: ./test.hlb:2:2: bad cast
------
```

This is probably a checker edge case, not sure.

### PR DESCRIPTION
**Pull request by hinshun**
There are 3 primary rules for binds listed below.
1. Option blocks do not have effect registers.
2. Binds must have access to effect registers.
3. `with` provides access to effect registers of the parent function.

The new parser.Match provides a much cleaner interface to match a sequence of
CST structures and avoids cluttering logic with interface casting code.

Fixes #162

Implementation notes:
- Sometimes, you don't care whether you're dealing with a `FuncDecl` or `FuncLit`. The common struct between them is `BlockStmt` but you lose access to scope/type information. In addition, `FuncLit` used with `WithOpt` infers additional type information from its parent `CallStmt` like `option` -> `option::run`, so I decided to add this semantic data into the `BlockStmt` instead. If you check the `BlockStmt.ObjType`, it will be correct and already inferred, as well as decoupled from `FuncDecl` or `FuncLit`.

### CODE REVIEW

File: checker/checker.go [hunk 1]
```diff
@@ -75,113 +84,127 @@ func (c *checker) Check(mod *parser.Module) error {
...
+		},
+		// Function literals propagate its scope to its children.
+		func(parentLit *parser.FuncLit, lit *parser.FuncLit) {
+			lit.Body.Scope = parentLit.Body.Scope
+			lit.Body.ObjType = parentLit.Body.ObjType
+		},
+		// WithOpt's function literals need to infer its secondary type from its
+		// parent call statement. For example, `run with option { ... }` has a
+		// `option` type function literal, but infers its type as `option::run`.
+		func(fun *parser.FuncDecl, call *parser.CallStmt, with *parser.WithOpt, lit *parser.FuncLit) {
```
[the remark this answers is not in the document]
Comment (hinshun) [reply checker/checker.go:116]: > First, doesn't with also need to supply the function register to non-literal options, eg with readonly?
Technically yes, but I don't think any non-literal options uses function registers atm. You're also not allowed to define binds when using a non-literal builtin in this short-hand style. Though I believe we can "fulfill" this by doing what you suggest later, which is to also have a `Lexical` pointer in `WithOpt`.

> the FuncDecl and FuncLit are sharing scope
The conceptual level is that the `FuncLit` should have a sub-scope deriving from `FuncDecl`'s scope. Implementation-wise, you can't define new decls within `FuncLit` atm, so there isn't any need to create a sub-scope. Thinking about this more, I think FuncLits shares scope / derives sub-scope, but they each have their own registers (except option func lits). The only exception is `with` that negotiates access between option func lits and the parent func's registers.

> would it make sense to memoize type information and lexical scope onto the WithOpt itself, so that later Matchers can use that to find the Lexical func?
I think this makes great sense. I need to find a way to pass down either `*FuncDecl` or `*WithOpt` to able to access the `Lexical` pointer though.

[the remark this answers is not in the document]
Comment (hinshun) [reply checker/checker.go:122]: FYI, renamed `Lexical` to `Closure` following the conversation on slack.
